src/main/java/uk/nhs/hee/trainee/details/service/NtnGenerator.java

```diff
@@ -0,0 +1,303 @@
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
+ * DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+package uk.nhs.hee.trainee.details.service;
+
+import java.time.LocalDate;
+import java.util.Comparator;
+import java.util.List;
+import java.util.ListIterator;
+import java.util.Objects;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.stereotype.Service;
+import uk.nhs.hee.trainee.details.model.Curriculum;
+import uk.nhs.hee.trainee.details.model.PersonalDetails;
+import uk.nhs.hee.trainee.details.model.ProgrammeMembership;
+import uk.nhs.hee.trainee.details.model.TraineeProfile;
+
+/**
+ * A service for handling trainee NTNs/DRNs.
+ */
+@Slf4j
+@Service
+public class NtnGenerator {
+
+  /**
+   * Populate the NTN for all programme memberships in the trainee profile.
+   *
+   * @param traineeProfile The trainee profile to populate with NTNs.
+   */
+  public void populateNtns(TraineeProfile traineeProfile) {
+    PersonalDetails personalDetails = traineeProfile.getPersonalDetails();
+
+    if (isExcluded(personalDetails)) {
+      return;
+    }
+
+    traineeProfile.getProgrammeMemberships().forEach(pm -> populateNtn(personalDetails, pm));
+  }
+
+  /**
+   * Populate the NTN for the given programme membership.
+   *
+   * @param personalDetails     The personal details to use for NTN generation.
+   * @param programmeMembership The programme membership to generate the NTN for.
+   */
+  private void populateNtn(PersonalDetails personalDetails,
+      ProgrammeMembership programmeMembership) {
+    log.info("Populating NTN for programme membership '{}'.", programmeMembership.getTisId());
+
+    if (isExcluded(programmeMembership)) {
+      return;
+    }
+
+    String parentOrganization = getParentOrganization(programmeMembership);
+    String specialtyConcat = getSpecialtyConcat(programmeMembership);
+    String referenceNumber = getReferenceNumber(personalDetails);
+    String suffix = getSuffix(programmeMembership);
+    String ntn = parentOrganization + "/" + specialtyConcat + "/" + referenceNumber + "/" + suffix;
+    programmeMembership.setNtn(ntn);
+    log.info("Populated NTN: {}.", ntn);
+  }
+
+  /**
+   * Get the parent organization for the given programme membership.
+   *
+   * @param programmeMembership The programme membership to calculate the parent organization for.
+   * @return The calculated parent organization.
+   */
+  private String getParentOrganization(ProgrammeMembership programmeMembership) {
+    String managingDeanery = programmeMembership.getManagingDeanery();
+    log.info("Calculating parent organization for managing deanery '{}'.", managingDeanery);
+
+    String parentOrganization = managingDeanery == null ? null : switch (managingDeanery) {
```

Comment:
😢 
Would this be more or less manageable as a lookup table I wonder (the complexity for South West notwithstanding)!

Reply:
It's really pretty, isn't it 😆

Reply:
I think South West will be moving to a single code, which may make a reference/lookup table easier to implement. But without having an easy way to manage it we'd end up making code changes (migration scripts) anyway, so we would end up with a more complex and less transparent solution for a little code tidiness 😞